Lf 3934 create separate number input with localization support

navDhammu opened the pull request:
**Description**

Created the following new components: 

**NumberInput**
- Handles numeric inputs and allows entering numbers in the specified locale.
- By default uses the locale defined in i18next, but provides a way to override with a custom locale through the locale prop. 
- Restricts any input that isn't a number or a decimal.
- Works with copy paste as long as the copied number is not formatted (only contains a decimal).
- UPDATE: This now works with currency as well. Simply pass in currencySymbol and the decimal places for that currency as props.

**InputBase**
- Serves as a reusable component from which other components can be made without having to copy paste the markup and styling from the existing Input.
- Most of the styles and jsx is copied over from Input, main difference is that it accepts a prop for the left section and right sections and leaves it up to the parent to decide what should be rendered in those sections.

**Usage with react hook form**
- Pass in name and control from RHF, there is no need to manually register it.
```
import { useForm } from 'react-hook-form';
import NumberInput from '../../Form/NumberInput';

function Component() {
  const { control } = useForm();
  return (
      <form>
        <NumberInput
          name='name'
          control={control}
             // other options
        />
      </form>
  );
}

```
Jira link: https://lite-farm.atlassian.net/browse/LF-3934

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Visually tested in Storybook as well as automated tests added with the Play function.

- [x] Passes test case
- [ ] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [x] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

navDhammu commented:
**Update:** 

Added some minor changes after I noticed an issue with using the allowDecimal and step prop together. Also added comments describing the component props.

navDhammu commented on packages/webapp/src/components/Form/InputBase/InputBaseField/index.tsx:
I've made some changes to the underlying html and css for the input here. As mentioned on slack there were some issues with icons overlapping each other due to absolute positioning inside the input (see below screenshot).

<img width="720" alt="Screen Shot 2024-01-12 at 12 33 12 PM" src="https://github.com/LiteFarmOrg/LiteFarm/assets/73000930/825bc9da-2f4b-40ba-8658-856e96f25c49">

What I did is create a wrapper div with display flex and style it like an input, remove all styles from the actual input, and place the left section, input, and right sections positioned inside wrapper div as flex children. This ensures that each element takes up the correct space and resolves the problem of overlapping icons.

SayakaOno reviewed (commented):
Thank you Nav for working on this!
I still have some files to go through but will leave questions I have so far!

One question, is it not possible to send a currency as a `leftSection`? I think it looks good...

navDhammu commented:
@SayakaOno Thank you for reviewing this!

For the currency question - yes its possible to send currency as leftSection but then you have to be careful with the other props you pass. For example if you pass in step, then the number will take the number of decimal places in step, and this might override the decimal places used for that currency (different currencies use different number of decimals). Also, for rounding, if you pass roundToDecimalPlaces={2} for currency, then a currency number that should be $2,392.10, will get rounded to $2,392.1. 

There is a currency formatter built in with Intl.NumberFormat and I thought of using that but it was making the current formatter logic more complicated to handle all different combination of props that can be passed.

navDhammu commented on packages/webapp/src/components/Form/InputBase/InputBaseLabel/index.tsx:
```diff
@@ -0,0 +1,47 @@
+
+  return (
+    <div className={styles.labelContainer}>
+      <Label style={{ position: 'absolute', bottom: 0 }}>
+        {props.label}
+        {props.optional && (
+          <Label sm className={styles.sm} style={{ marginLeft: '4px' }}>
```
Yes its possible - I just copied that over from the Input file but I can make the changes if needed.

SayakaOno replied:
That would be great, thank you! We are trying to move away from inline CSS and makeStyles (from material UI) and use scss files.

navDhammu commented on packages/webapp/src/components/Form/InputBase/InputBaseField/index.tsx:
```diff
@@ -0,0 +1,54 @@
+      )}
+    >
+      {props.leftSection && (
+        <div className={styles.inputSection} data-section="left">
```
Ya I wanted to add padding on left and right sections, not sure why I didn't just use classnames now that I think of it.

SayakaOno replied:
Would you be able to update these to use classNames?
Just to make things consistent!

navDhammu commented on packages/webapp/src/components/Form/NumberInput/NumberInput.tsx:
No it shouldn't be possible to enter negative numbers.

SayakaOno replied:
Would it be challening to handle negative numbers??

navDhammu replied:
I don't think it will be challenging, do we need to handle negative numbers in the app?

SayakaOno replied:
I don't think we currently have inputs that allow negative numbers, but I would try not to limit the capability of  the native input and also consider possible future uses.
But I'm not sure! @antsgar what do you think?

SayakaOno commented:
@navDhammu Oh that's the complexity you were referring to! I never knew the input component had such complicated logic 😅 Thank you for providing the detail!

SayakaOno commented on packages/webapp/src/components/Form/NumberInput/utils.ts:
```diff
@@ -0,0 +1,31 @@
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+export function countDecimalPlaces(number: number) {
```
Is this the best way to count decimal places? I was just trying to figure out what this does and happened to find that the function returns `3` for the input `1.11`.

navDhammu replied:
Oh I see what's happening there. Thank you for catching this! I will try to explain what its doing and what I did wrong.

This function is taking the number passed in and multiplying by e (power of 10) on each iteration and moving the decimal to the right by one place until it reaches the end. Using the % (modulo) operator, it checks if there is a remainder and stops when (number*e % 1) = 0, indicating that there are no more decimal places.

For example, if you pass in 1.11:

**First iternation:**
e=1
decimalPlaces=0
 (1.11 * 1) % 1 = 0.11

**Second iternation:**
e=10
decimalPlaces=1
 (1.11 * 10) % 1 = 0.1

**Third iternation:**
e=100
decimalPlaces=2
 (1.11 * 100) % 1 = 0

The thing I overlooked is that in JavaScript, sometimes when you multiply a decimal number it doesn't give the result with exact precision. Like 1.11*100 = 111.00000000001, not 111, so in the example above, third iteration is not true and the loop continues.

navDhammu replied:
I fixed it by rounding to a whole number (Math.Round(number*e)) and dividing by e, then checking if it equals the original number. Also added test cases for it and also for the clamp function.

SayakaOno reviewed (commented):
Thank you so much again for working on this, and I'm sorry that I couldn't finish reviewing sooner.
Code and tests look great!

I think it's assumed that inputs can be operated using the keyboard's up and down arrows. I couldn't catch the code disabling this feature while reviewing the code, but was that intentional?

navDhammu commented on packages/webapp/src/components/Form/NumberInput/NumberInput.tsx:
```diff
@@ -0,0 +1,253 @@
+  const [isFocused, setIsFocused] = useState(false);
+  const [isEditedAfterFocus, setIsEditedAfterFocus] = useState(false);
+  const initialValueRef = useRef(propValue);
+  const showStepper = typeof step === 'number' && step > 0;
```
We don't need to check typeof step but since step is optional it can be undefined so we have to check if it exists. One alternative is to do: `!!step && step > 0`. Or we can also default step to 0 and then it just becomes 
`const showStepper = step > 0`

navDhammu commented:
@SayakaOno Regarding the keyboard up and down arrows. We're using the text input instead of the native number input and so that doesn't have default up and down arrow buttons and keyboard behaviour. But if that's important then it can be added in, I just wanted to keep it as simple as possible for now.

navDhammu commented on packages/webapp/src/components/Form/NumberInput/NumberInput.tsx:
```diff
@@ -0,0 +1,253 @@
+   * If allowDecimal is false, then step is rounded down to the nearest whole number.
+   */
+  step?: number;
+  /**
+   * Max value of input. Only applicable if used with step
+   */
+  max?: number;
+  /**
+   * Min value of input. Only applicable if used with step
+   */
+  min?: number;
```
The purpose of max and min here is to disable the up/down buttons to prevent the user from going past the min or max. So it doesn't make sense to use without step. But for validation purposes we can still pass min or max to react hook form, like: 

```
<NumberInputRHF 
   name='some-name'
   rules={   
      max: {
        value: 20,
        message: 'error message'
    },    
     min: {
       value: 5,
       message: 'error message'
    }
/>
```

SayakaOno commented:
@navDhammu Thank you for clarifying! I think it should work like the native number input, but I will leave the decision to @antsgar!

navDhammu commented:
@antsgar @SayakaOno I've made the following new changes:
- Added support for the arrow up/down keys
- Added support for currency by making some changes to the number formatting options. Now we can simply pass `currencySymbol` as `leftSection` along with `decimalDigits` to correctly display the currency. So creating a separate currency component isn't needed.
- Some refactoring in the main NumberInput component
- Minor changes as suggested by @SayakaOno 

I think this component is now almost finished, I'm just trying to work out a few minor things like optimizations and better integration with react hook form. 

Meanwhile if I can get some feedback that would be great, thanks!

navDhammu commented:
This component should now be ready for a final review. @antsgar @SayakaOno

If this can be merged in then it can help in creating the [component for sex detail](https://lite-farm.atlassian.net/browse/LF-4154), as we can simply call the useNumberInput hook and render the custom stepper buttons with left and right sections utilizing InputBase

SayakaOno commented on packages/webapp/src/components/Form/InputBase/InputBaseField/styles.module.scss:
```diff
@@ -0,0 +1,78 @@
+}
+
+.inputError {
+  border-color: var(--error);
```
I think the border also needs to be this colour.

navDhammu replied:
Do you mean it should be red when focused? I looked at the current Input and it doesn't have the red border when it's focused in error state.

antsgar reviewed (dismissed):
Left a small comment but this looks great, thank you for working on it!! Awesome work 🙌

navDhammu commented on packages/webapp/src/components/Form/InputBase/InputBaseField/styles.module.scss:
```diff
@@ -0,0 +1,78 @@
+  border: 1px solid var(--grey400);
+  box-sizing: border-box;
+  border-radius: 4px;
+  height: 40px;
```
I think the new inputs in figma now have 40px height?  Are we keeping the old designs for inputs or changing to new ones going forward?

navDhammu commented on packages/webapp/src/stories/Form/NumberInput/NumberInput.stories.tsx:
```diff
@@ -0,0 +1,371 @@
+export const WithOptionalLabel: Story = {
+  args: {
+    label: 'A label',
+    optional: true,
```
It was happening because placeholder wasn't specified as a prop, I did a bit of refactoring with the props in InputBase so it should be fixed now.

navDhammu commented on packages/webapp/src/components/Form/NumberInput/useNumberInput.ts:
```diff
@@ -0,0 +1,197 @@
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const stepValue = allowDecimal ? step : Math.round(step);
+  const showStepper = stepValue > 0;
```
Thanks for pointing this out, I changed it to accept a `showStepper` prop which is false by default so the stepper is hidden. And the keyboard up/down works in all cases.

navDhammu commented on packages/webapp/src/components/Form/InputBase/InputBaseField/index.tsx:
```diff
@@ -0,0 +1,52 @@
+import clsx from 'clsx';
+
+type InputBaseFieldProps = Pick<InputBaseProps, 'leftSection' | 'rightSection'> & {
+  crossIcon?: ReactElement;
```
Would something like `resetIcon` be better? Because the input can have an initial value and in that case clicking this icon wouldn't clear the entire value but reset it to its initial state.

SayakaOno reviewed (approved):
Thank you so much!!!
